File: 02/02.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func SplitByLines(bytes []byte) []string {
	return strings.Split(strings.Trim(string(bytes[:]), "\n"), "\n")
}

func Less(a, b int) bool {
	return a < b
}
func More(a, b int) bool {
	return a > b
}
func P(f func(int, int) bool, a, b int) bool {
	return f(a, b) && Abs(a-b) <= 3
}
func Abs(x int) int {
	if x < 0 {
		x = -x
	}
	return x
}
func Safe1(l []int) bool {
	if l[0] == l[1] {
		return false
	}
	f := Less
	if l[0] > l[1] {
		f = More
	}
	j := 1
	for j < len(l) && P(f, l[j-1], l[j]) {
		j++
	}
	r := j < len(l)
	return !r
}
func Safe2(l []int) bool {
	cl := 0
	cm := 0
	cq := 0
	il := -1
	im := -1
	iq := -1
	for i := 1; i < len(l); i++ {
		a := l[i-1]
		b := l[i]
		if a >= b {
			cl++
			if il == -1 {
				il = i
			}
		}
		if a <= b {
			cm++
			if im == -1 {
				im = i
			}
		}
		if Abs(a-b) > 3 {
			cq++
			if iq == -1 {
				iq = i
			}
		}
	}
	if cq == 0 && (cm == 0 || cl == 0) {
		return true
	}
	if cq > 2 || cl > 2 && cm > 2 {
		return false
	}
	if cq == 2 {
		return Safe1(Skip(l, iq))
	}
	if iq != -1 {
		return Safe1(Skip(l, iq)) || Safe1(Skip(l, iq-1))
	}
	if 0 < cl && cl <= 2 {
		return Safe1(Skip(l, il)) || Safe1(Skip(l, il-1))
	}
	if 0 < cm && cm <= 2 {
		return Safe1(Skip(l, im)) || Safe1(Skip(l, im-1))
	}
	return false
}
func Skip(l []int, i int) []int {
	n := make([]int, len(l)-1)
	k := 0
	for j, x := range l {
		if j != i {
			n[k] = x
			k++
		}
	}
	return n
}
func main() {
	bytes, err := os.ReadFile("./input.txt")
	if err != nil {
		panic(err)
	}
	lines := SplitByLines(bytes)
	data := make([][]int, len(lines))
	for i, l := range lines {
		nums := strings.Split(l, " ")
		data[i] = make([]int, len(nums))
		for j, c := range nums {
			if k, e := strconv.Atoi(c); e == nil {
				data[i][j] = k
			}
		}
	}
	s := 0
	for _, l := range data {
		if Safe1(l) {
			s++
		}
	}
	fmt.Println(s)
	s = 0
	for _, l := range data {
		if Safe2(l) {
			s++
		}
	}
	fmt.Println(s)
}

File: 02/02.py



data = open('input.txt').read().strip().split("\n")
data = [list(map(int, x.strip().split(" "))) for x in data]

def less(a, b):
    return a<b
def more(a, b):
    return a>b
q = lambda a, b: abs(a-b)<=3

def safe1(l, f=None):
    if not f:
        if l[0]<l[1]:
            f = less
        elif l[0]>l[1]:
            f = more
        else:
            return False
    return all(f(a, b) and q(a, b) for a, b in zip(l, l[1:]))


def safe2(l):
    for i in range(len(l)):
        t = l[:]
        del t[i]
        if safe1(t):
            return True
    return False

def safe3(l):
    cl = 0
    cm = 0
    cq = 0
    il, im, iq = None, None, None
    for i, (a, b) in enumerate(zip(l, l[1:]), 1):
        if a>=b:
            cl += 1
            if not il:
                il = i
        if a<=b:
            cm += 1
            if not im:
                im = i
        if not q(a, b):
            cq += 1
            if not iq:
                iq = i
    if cq==0 and (cm==0 or cl==0):
        return True
    if cq>2 or cl>2 and cm>2:
        return False
    if cq==2:
        del l[iq]
        return safe1(l)
    if iq:
        t = l[:]
        del t[iq-1]
        del l[iq]
        return safe1(l) or safe1(t)
    if 0<cl<=2:
        t = l[:]
        del t[il-1]
        del l[il]
        return safe1(l) or safe1(t)
    if 0<cm<=2:
        t = l[:]
        del t[im-1]
        del l[im]
        return safe1(l) or safe1(t)
    return False

try:
    import pytest
        
    @pytest.mark.parametrize("l", data)
    def test_safe2_vs_safe3(l):
        assert safe2(l)==safe3(l)

    def test_len():
        assert all(len(x)>3 for x in data)
except:
    pass

if __name__ == "__main__":
    print(sum(1 for x in data if safe1(x)))
    print(sum(1 for x in data if safe3(x)))
